### PR TITLE
CRYPTO-96: OpenSSL Random implementation silently falls back to Java

### PR DESCRIPTION
**Pull request by sundapeng**
https://issues.apache.org/jira/browse/CRYPTO-96

### CODE REVIEW

File: src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java [hunk 1]
```diff
@@ -44,46 +43,25 @@
...
     // N.B. this class is not public/protected so does not appear in the main Javadoc
     // Please ensure that property use is documented in the enum CryptoRandomFactory.RandomProvider
-    public OpenSslCryptoRandom(Properties props)
-            throws NoSuchAlgorithmException {
-        //fallback needs to be initialized here in any case cause even if
-        //nativeEnabled is true OpenSslCryptoRandomNative.nextRandBytes may fail
-        fallback = new JavaCryptoRandom(props);
+    public OpenSslCryptoRandom(Properties props) {
+        // Generates one byte to verify Openssl random
+        next(1);
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java]: Surely one needs to check the return value here?

File: src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java [hunk 2]
```diff
@@ -93,9 +71,7 @@ public OpenSslCryptoRandom(Properties props)
      */
     @Override
     public void nextBytes(byte[] bytes) {
-        if (!nativeEnabled || !OpenSslCryptoRandomNative.nextRandBytes(bytes)) {
-            fallback.nextBytes(bytes);
-        }
+        OpenSslCryptoRandomNative.nextRandBytes(bytes);
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java]: Again, the return value needs to be checked

File: src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java [hunk 3]
```diff
@@ -140,8 +116,5 @@ final protected int next(int numBits) {
      */
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java]: Javadoc needs adjusting now that the code has changed

File: src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java [hunk 4]
```diff
@@ -44,46 +43,25 @@
...
-    private static final boolean nativeEnabled;
-
     static {
-        boolean opensslLoaded = false;
         if (Crypto.isNativeCodeLoaded()) {
             try {
                 OpenSslCryptoRandomNative.initSR();
-                opensslLoaded = true;
             } catch (Exception t) {// NOPMD
             }
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java]: What if the native code is not loaded, or initSR fails?
This should surely be noted here in a final boolean and used in the ctor and methods?

File: src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java [hunk 5]
```diff
@@ -44,46 +43,25 @@
...
-    }
-
-    /**
-     * Judges whether loading native library successfully.
-     *
-     * @return true if loading library successfully.
-     */
-    public static boolean isNativeCodeLoaded() {
-        return nativeEnabled;
     }
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java]: Surely this method is still needed?

**Comment by sundapeng** [comment]
Thank @sebbASF ，patch updated

**Comment by sebbASF** [comment]
Thanks!

I've been looking into this further, and I think we could/should provide more specific feedback for failures.
For example the static block could save the exception when initSR fails.
Also nextBytes does not need to check nativeEnabled.
See attached patch for sample implementation.
[crypto96-patch.txt](https://github.com/apache/commons-crypto/files/349504/crypto96-patch.txt)

**Comment by sundapeng** [comment]
Thank @sebbASF for your comment. I will commit it.

**Pull request closed without merging**